### PR TITLE
Refactor the application to reduce code duplication and improve overall readability

### PR DESCRIPTION
**Pull request by Yantrio**
Part of #59 

This PR Refactors quite a lot of the repo to clean things up. It does the following:

- Introduces one centralized `ConfigBuilder` to be used by all lambdas
- Config object now contains clients for aws services (lambda etc)
- Introduce some wrapper classes for non-simple services (secrets, dynamodb cache for provider versions)
- Reworks a lot of the code to remove duplicated logic

### CODE REVIEW

File: src/internal/providers/providercache/store.go [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+	//tableName := os.Getenv("PROVIDER_VERSIONS_TABLE_NAME")
+	//if tableName == "" {
+	//	panic(fmt.Errorf("missing environment variable PROVIDER_VERSIONS_TABLE_NAME"))
+	//}
+	//
+	//awsConfig, err := awsconfig.LoadDefaultConfig(ctx, awsconfig.WithRegion(os.Getenv("AWS_REGION")))
+	//if err != nil {
+	//	return nil, fmt.Errorf("could not load AWS configuration: %w", err)
+	//}
+
```
Comment (RLRabinowitz) [inline src/internal/providers/providercache/store.go]: Leftover comments

File: src/internal/providers/providercache/store.go [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+func (p *Handler) Store(ctx context.Context, providerNamespace string, providerType string, versions []providers.Version) error {
+	provider := fmt.Sprintf("%s/%s", providerNamespace, providerType)
+	//
+	//awsConfig, err := awsconfig.LoadDefaultConfig(ctx, awsconfig.WithRegion(os.Getenv("AWS_REGION")))
+	//if err != nil {
+	//	return fmt.Errorf("could not load AWS configuration: %w", err)
+	//}
+	//
+	//ddbClient := dynamodb.NewFromConfig(awsConfig)
```
Comment (RLRabinowitz) [inline src/internal/providers/providercache/store.go]: Same, leftover comments

File: src/internal/providers/providercache/fetch.go [hunk 3]
```diff
@@ -0,0 +1,33 @@
+package providercache
+
+import (
+	"context"
+	"github.com/aws/aws-sdk-go-v2/feature/dynamodb/attributevalue"
+	"github.com/aws/aws-sdk-go-v2/service/dynamodb"
+	"github.com/aws/aws-sdk-go-v2/service/dynamodb/types"
+)
+
+func (p *Handler) GetItem(ctx context.Context, key string) (*VersionListingItem, error) {
```
Comment (RLRabinowitz) [inline src/internal/providers/providercache/fetch.go:10]: `GetItem` and `Store` should match in their arguments, regarding the key. Here we get the `key` is input, but in `Store` we build it from `namespace`+`type`

File: src/lambda/api/providerVersions.go [hunk 4]
```diff
@@ -50,101 +48,85 @@ func listProviderVersions(config Config) LambdaFunc {
...
 		}
 		if document != nil {
-			// TODO: If the document is more than an hour old, invoke the lambda to update it.
-
-			fmt.Printf("Found document with %d versions\n", len(document.Versions))
+			fmt.Printf("Found document with %d versions, last_updated: %s\n", len(document.Versions), document.LastUpdated.String())
+			// if the document is within our accepted age range, return it
+			if document.LastUpdated.After(time.Now().Add(-providerCacheAge)) {
+				fmt.Printf("Document is recent enough, returning it\n")
+			}
```
Comment (RLRabinowitz) [inline src/lambda/api/providerVersions.go]: The `if` statement is wrong here, unless you add a return statement here `return foundDocumentResponse(document)`. Consider either doing that, or inverting the condition

File: src/lambda/api/providerVersions.go [hunk 5]
```diff
@@ -50,101 +48,85 @@ func listProviderVersions(config Config) LambdaFunc {
...
+
+			// else the document is too old, we should update the cache
+			fmt.Printf("Document is too old, invoking lambda to update dynamodb\n")
+			err = triggerPopulateProviderVersions(ctx, config, effectiveNamespace, params.Type)
```
Comment (RLRabinowitz) [inline src/lambda/api/providerVersions.go]: I think it's a bit confusing how we handle the cache of the providers that are included in the `providerRedirects` configuration.

We always query the cache by `effectiveNamespace`, and for storing we send the `effectiveNamespace` to the `populate_provider_versions` lambda, which considers it the `namespace`.
It's nice that `populate_provider_versions` is not aware of the provider redirect logic at all, though.

Maybe a better approach would be to store and query the cache by the provider **without** using the effective namespace at all? WDYT?

File: src/lambda/populate_provider_versions/handler.go [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+			}
+
+			fmt.Printf("Repo %s/%s exists\n", e.Namespace, repoName)
+
+			v, err := providers.GetVersions(tracedCtx, config.RawGithubv4Client, e.Namespace, repoName)
+			if err != nil {
+				return fmt.Errorf("failed to get versions: %w", err)
+			}
+
+			fmt.Printf("Found %d versions\n", len(v))
```
Comment (RLRabinowitz) [inline src/lambda/populate_provider_versions/handler.go:62]: You can probably share this whole logic with the `api` lambda. Not critical, though

File: src/lambda/api/providerVersions.go (hunk 4 above)
Comment (Yantrio) [reply src/lambda/api/providerVersions.go]: This is what I get for working late! I'll fix this small oversight, thanks

File: src/lambda/api/providerVersions.go (hunk 5 above)
Comment (Yantrio) [reply src/lambda/api/providerVersions.go]: Honestly I think I prefer the idea of the `populate_provider_versions` not knowing about the redirection logic, this means that we are storing a 1:1 relationship between github repos and provider releases, and in the future we can easily add more redirects later on if needed.

Comment (RLRabinowitz) [reply src/lambda/api/providerVersions.go]: WDYT about this approach then?

> Maybe a better approach would be to store and query the cache by the provider without using the effective namespace at all?

Not too important though. Feel free to resolve this conversation

File: src/internal/providers/providercache/fetch.go (hunk 3 above)
Comment (Yantrio) [reply src/internal/providers/providercache/fetch.go:10]: I'll move both to use `key` to keep it simple :D

File: src/internal/config/config.go [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+}
+
+// BuildConfig will build a configuration object for the application. This
+// includes loading secrets from AWS Secrets Manager, and configuring the
+// AWS SDK.
+func (c ConfigBuilder) BuildConfig(ctx context.Context, xraySegmentName string) (config *Config, err error) {
+	if err = xray.Configure(xray.Config{ServiceVersion: "1.2.3"}); err != nil {
+		err = fmt.Errorf("could not configure X-Ray: %w", err)
+		return
+	}
```
Comment (eranelbaz) [inline src/internal/config/config.go:55]: why do `err` needs to be assigned if you `return` nothing?

File: src/internal/secrets/secretsmanager.go [hunk 8]
```diff
@@ -4,29 +4,43 @@ import (
...
+	value, err := s.client.GetSecretValue(ctx, &secretsmanager.GetSecretValueInput{
 		SecretId: aws.String(secretName),
 	})
 	if err != nil {
 		return "", err
 	}
 	return *value.SecretString, nil
 }
+
+func (s *Handler) GetValueFromEnvVar(ctx context.Context, envVarName string) (string, error) {
```
Comment (eranelbaz) [inline src/internal/secrets/secretsmanager.go]: why this one in secret manager if that's env var?

File: src/lambda/populate_provider_versions/handler.go (hunk 6 above)
Comment (Yantrio) [reply src/lambda/populate_provider_versions/handler.go:62]: I agree, I think I want to try and refactor out some of the repository exists stuff to a middleware layer but I couldn't find a nice way to do it without some hacky code. I'll revisit this later on

File: src/internal/config/config.go (hunk 7 above)
Comment (Yantrio) [reply src/internal/config/config.go:55]: this is to return the named response item `err error`

File: src/internal/secrets/secretsmanager.go (hunk 8 above)
Comment (Yantrio) [reply src/internal/secrets/secretsmanager.go]: It's getting a secret value from secretsmanager by using the env var name to figure out what key to fetch.

I couldn't find a nice way to name this function. I'll try and think of something better.

Comment (Yantrio) [reply src/internal/secrets/secretsmanager.go]: WDYT about `GetSecretValueFromEnvReference` ?

File: src/lambda/api/providerVersions.go (hunk 5 above)
Comment (Yantrio) [reply src/lambda/api/providerVersions.go]: I have a feeling that there may be some clashes. especially as multiple forks of providers happen in the future. I don't think it's safe to make the assumption that provider name is all we need.

File: src/internal/secrets/secretsmanager.go (hunk 8 above)
Comment (eranelbaz) [reply src/internal/secrets/secretsmanager.go]: Sounds good

File: src/lambda/api/providerVersions.go [hunk 9]
```diff
@@ -33,118 +33,85 @@ type ListProviderVersionsResponse struct {
...
 		repoName := providers.GetRepoName(params.Type)
 
-		// check the repo exists
-		exists, err := github.RepositoryExists(ctx, config.ManagedGithubClient, effectiveNamespace, repoName)
-		if err != nil {
-			return events.APIGatewayProxyResponse{StatusCode: 500}, err
-		}
-		if !exists {
+		if exists, err := github.RepositoryExists(ctx, config.ManagedGithubClient, effectiveNamespace, repoName); !exists {
+			if err != nil {
```
Comment (eranelbaz) [inline src/lambda/api/providerVersions.go:43]: Maybe worth logging? 
So we'll know if we are missing something somehow?

File: src/lambda/api/providerVersions.go [hunk 10]
```diff
@@ -33,118 +33,85 @@ type ListProviderVersionsResponse struct {
...
+	if err := triggerPopulateProviderVersions(ctx, config, namespace, providerType); err != nil {
+		fmt.Printf("Error triggering lambda to update dynamodb: %s\n", err.Error())
 	}
+
+	return versionsResponse(document.Versions, nil)
 }
 
-func foundDocumentResponse(document *versions_cache.ProviderVersionListingItem) (events.APIGatewayProxyResponse, error) {
+func fetchFromGithub(ctx context.Context, config config.Config, namespace, repoName string) (events.APIGatewayProxyResponse, error) {
+	fmt.Printf("Document not found in dynamodb, invoking lambda and loading from github\n")
```
Comment (eranelbaz) [inline src/lambda/api/providerVersions.go]: Worth logging which

**Review by eranelbaz — APPROVED** [review]
Some logging comments 
lgtm

File: src/lambda/api/providerVersions.go (hunk 9 above)
Comment (Yantrio) [reply src/lambda/api/providerVersions.go:43]: yes, I'll add logging here. especially now X-ray is inconsistent :)

**Review by RLRabinowitz — APPROVED** [review]

**Pull request merged**